README.md
Torque
==

[![Build Status](https://travis-ci.org/CartoDB/torque.svg?branch=master)](https://travis-ci.org/CartoDB/torque)

Render big, timeseries data in the client. Uses CartoDB to generate a datacube format. For a brief introduction to the format and methods, see our [presentation slides](http://gijs.github.com/images/cartodb_datacubes.pdf)

CartoDB users can use this code right away. Specify your username/tablename and datecolumn to get mapping time immediatley.

Have fun!

* Play with your own data: http://cartodb.github.com/torque/
* The Guardian's Data Blog write-up of Royal Navy ships in WWI: http://www.guardian.co.uk/news/datablog/interactive/2012/oct/01/first-world-war-royal-navy-ships-mapped
* A car in Nürburgring track: http://cartodb.github.com/torque/examples/car.html

## Torque library reference

Torque lets you render big, timeseries or categorical data in the client. This is useful for many modern applications that need more than just a static map. Early versions of Torque have been used to visualize human movement, Twitter activity, biodiversity data, and many more large-scale datasets.

The library uses CartoDB to generate a [layercube](https://github.com/CartoDB/tilecubes/blob/master/2.0/spec.md) format. For a brief introduction to the format and methods, see our [presentation slides](http://gijs.github.com/images/cartodb_datacubes.pdf). If you are not using CartoDB to host your data, you can modify the input parameters to point toward any layercube service.

### Getting started

The simplest way to use a visualization with Torque is...

#### Create a simple Torque visualization

```javascript
  ...
  <body>
    <div id="map"></div>
    ...
    <script>
      // define the torque layer style using cartocss
      var CARTOCSS = [
          'Map {',
          '-torque-time-attribute: "date";',
          '-torque-aggregation-function: "count(cartodb_id)";',
          '-torque-frame-count: 760;',
          '-torque-animation-duration: 15;',
          '-torque-resolution: 2',
          '}',
          '#layer {',
          '  marker-width: 3;',
          '  marker-fill-opacity: 0.8;',
          '  marker-fill: #FEE391; ',
          '  comp-op: "lighten";',
          '  [value > 2] { marker-fill: #FEC44F; }',
          '  [value > 3] { marker-fill: #FE9929; }',
          '  [value > 4] { marker-fill: #EC7014; }',
          '  [value > 5] { marker-fill: #CC4C02; }',
          '  [value > 6] { marker-fill: #993404; }',
          '  [value > 7] { marker-fill: #662506; }',
          '  [frame-offset = 1] { marker-width: 10; marker-fill-opacity: 0.05;}',
          '  [frame-offset = 2] { marker-width: 15; marker-fill-opacity: 0.02;}',
          '}'
      ].join('\n');

        
      var map = new L.Map('map', {
        zoomControl: true,
        center: [40, 0],
        zoom: 3
      });

      L.tileLayer('http://{s}.api.cartocdn.com/base-dark/{z}/{x}/{y}.png', {
        attribution: 'CartoDB'
      }).addTo(map);

      var torqueLayer = new L.TorqueLayer({
        user       : 'viz2',
        table      : 'ow',
        cartocss: CARTOCSS
      });
      torqueLayer.addTo(map);
      torqueLayer.play()
    </script>
  </body>
```

[Grab the complete example source code](https://github.com/CartoDB/torque/blob/master/examples/navy_leaflet.html)
<div class="margin20"></div>

### API
[see reference](https://github.com/CartoDB/torque/blob/master/doc/API.md)



NEWS
3.1.2 (2019-mm-dd)


3.1.1 (2019-06-24)
    - Fixed reading undefined layer #306

3.1.0 (2019-04-02)
    - Upgrade canvas to version 2.4.1

3.0.0 (2019-03-29)
    - Drop support for Node.js 6 and 8
    - Handle exception: Image given has not completed loading

2.17.1 (2019-01-23)
    - Upgrade deps:
        - turbo-carto@0.21.1
    - Upgrade devel deps:
        - browserify@16.2.3
        - qunitz@0.7.7
        - request@2.88.0
        - uglify-js@3.4.9

2.17.0 (2018-11-21)
    - Use our current fork of mapnik while testing
    - Add support for Node 8 & 10
    - Support package-lock.json file

v2.16.5
    -

v2.16.4
    -

2.16.2
    - Re-publish version

2.16.1
    - Remove width trick to use clear rect #280
    - Map option reference broken gmaps #279

2.16.0
    - Updated Turbo-carto dependency to version 0.19.0.

2.15.1
    - Fixed ready event then rendering more than one step

2.15.0
    - Allow to render several steps at the same time (#247)

2.14.0
    - Allow custom animator range (#243)

2.13.0
    - Pass extra_params to server (#234)

2.12.0
    - Added TileJSON provider
    - Added getValues to fetch values for step
    - Improved reference/api documentation
    - Removed out of date examples

2.11.4
    - Static torque maps now draw correctly when unhiding them
    - Fixed animation always playing when rescaling
    - Added animator tests
2.11.3
    - Limited sprite radius to a maximum of 255px
    - Fixed custom marker-file functionality in Firefox
    - Fixed inertia mismatch of Torque's canvas layer
    - Added opacity get/setter to canvas layer in Google Maps
    - Fixed issue with `_super` being removed from listeners
2.11.2
    - Added error handling to Torque
    - Fixed 'remove' event triggering when removing a Torque layer.
    - Value is now capped at 255 instead of performing modulo on higher values.
2.11.1
    - Adapts usage of marker-opacity to Mapnik's functionality
    - Added getValueForBBox to TorqueLayer's API

2.11.0
    - Do not fix values in the edge (#147)
    - Windshaft provider accepts an optional data type for coordinates (#149)
    - Acceptance tests
    - renderTile accepts a callback to be called when rendering finishes
    - tile rendering deferred until all assets are loaded
    - sprite rendering now scales source to marker dimensions
    - Filters accept a canvas class

2.10.1
    - Adjusted point position taking resolution into account
    - Changed loop: false to pause animation at last frame
    - On torque-frame-count: 1, always pause for better performance
    - Fixes getTimeSpan

2.10.0
    - Adds timetostep method to torque layer
    - Function qualifyURL can be provided to not rely on document's one

2.9.0
    - Added heatmaps
    - Added image-filters
    - Solved problem of CSS not updating the vis until reloaded
    - Improved out of bounds correction function
    - Fixed image-asset sizing
    - Fixed items to load counter being sometimes negative

2.8.0
    - Supports use of external image markers via marker-file
    - Fixes Torque tiles not rendering outside normal lat/lng ranges.
    - Fixes marker image default pivot point
    - Fixes ghost frame being added to the end of animation
    - Moved to browserify for compatibility with Node.js
2.7.1
    - Fix renderer not rendering marker-lines

2.7.0
    - Adds option to not fetch map when instantiating a windshaft provider (#74)

2.6.1
    - Fix marker fill opacity (#72)

2.6.0
    - Published using browserify

2.5.xx
    - Added support for several auth_token param
    - Added support for cdn_url in winshaft provider
    - Added option to jsonp to set custom callback name
    - Added no_cdn option to windshaft provider
    - Added getActivePointsBBox method
    - Added renderer cache when the map is not animated
    - Added interaction method
    - Added stat_tag in the map creation url for windshaft provider
    - Added invalidate method to L.TorqueLayer

2.4.01
    - fixed rectangle anchor and size #42
    - fixed on method in torque.Event
    - fixed problem updating steps when sql change in windshaft provider

2.4.00 - May/22/2014
    - Changed date parsing for windshaft provider
    - Optimized tile loading
    - Fixed cartocss update when windshaft.provider was being used
    - Fixed windshaft provider with https
    - Generate cartocss in windshaft provider (#37)
    - Added zoom animation for leaflet
    - Added metrics to track memory and time
    - Pause on zoom
    - Fixed sprite rendering when line width is a float

2.3.00 - Feb/24/2014
-----
    - Added 'get attribution' function to Leaflet canvas layer.
    - Added zIndex support to leaflet torque layer
    - Added CartoDB Maps API data provider (used by default)
    - Fixed rendering when CartoCSS had zoom conditions (#27)

2.2.00 - Nov/26/2013
-----
    - added torque cumulative support #21

2.1.00 - 11/11/2013
-----
    - added support to custom SQL query
    - removed is_time option, it's not fetch from SQL-API
    - steps is now the upper limit for steps
    - multiple fixes


src/grid_layer.js
/*
 ====================
 this class renders tile data in a given time
 ====================
 */


function TimePlayer(min_date, end, step, options) {
    this.time = 0;
    this.step = step;
    this.CAP_UNIT = end;
    this.MIN_DATE = min_date;
    this.MAX_UNITS = options.steps + 2;
    this.MAX_VALUE = 0;
    this.MAX_VALUE_LOG = 0;
    this.BASE_UNIT = 0;
    this.canvas_setup = this.get_time_data;
    this.render = this.render_time;
    this.cells = [];
    this.table = options.table;
    this.user = options.user;
    this.t_column = options.column;
    this.resolution = options.resolution;
    this.countby = options.countby
    this.base_url = 'http://' + this.user + '.cartodb.com/api/v2/sql';
    this.options = options;
}

TimePlayer.prototype = new CanvasTileLayer();

/**
 * change time, t is the month (integer)
 */
TimePlayer.prototype.set_time = function (t) {
    if (this.time != (t >> 0)) {
        this.time = t;
        this.redraw();
    }
};
TimePlayer.prototype.reset_max_value = function () {
    this.MAX_VALUE = 0;
    this.MAX_VALUE_LOG = 0;
};
/**
 * change table where the data is choosen
 */
TimePlayer.prototype.set_table = function (table, size) {
    if (this.table === table) {
        return; // nothing to do
    }
    this.table = table;
    this.pixel_size = size;
    this.recreate();
    this.redraw();
};

/**
 * private
 */

// get data from cartodb
TimePlayer.prototype.sql = function (sql, callback) {
    var self = this;
    var uri  = this.base_url + "?q=" + encodeURIComponent(sql);
    //uri += '&t=' + Date.now() + Math.random();
    $.getJSON(uri, function (data) {
        callback(data);
    });
};

var originShift = 2 * Math.PI * 6378137 / 2.0;
var initialResolution = 2 * Math.PI * 6378137 / 256.0;
function meterToPixels(mx, my, zoom) {
    var res = initialResolution / (1 << zoom);
    var px = (mx + originShift) / res;
    var py = (my + originShift) / res;
    return [px, py];
}

// precache data to render fast
TimePlayer.prototype.pre_cache_months = function (rows, coord, zoom) {
    var row;
    var xcoords;
    var ycoords;
    var values;
    if (typeof(ArrayBuffer) !== undefined) {
        xcoords = new Uint8Array(new ArrayBuffer(rows.length));
        ycoords = new Uint8Array(new ArrayBuffer(rows.length));
        values = new Uint8Array(new ArrayBuffer(rows.length * this.MAX_UNITS));// 256 months
    } else {
        // fallback
        xcoords = [];
        ycoords = [];
        values = [];
        // array buffer set by default to 0        
        for (var i = 0; i < rows.length * this.MAX_UNITS; ++i) {
            values[i] = 0;
        }
    }
    // base tile x, y
    var tile_base_x = coord.x * 256;
    var tile_base_y = coord.y * 256;
    var total_pixels = 256 << zoom;
    for (var i in rows) {
        row = rows[i];
        pixels = meterToPixels(row.x, row.y, zoom);
        pixels[1] = total_pixels - pixels[1];
        xcoords[i] = pixels[0];
        ycoords[i] = pixels[1];
        var base_idx = i * this.MAX_UNITS;
        //def[row.sd[0]] = row.se[0];
        for (var j = 0; j < row.dates.length; ++j) {
            values[base_idx + row.dates[j]] = row.vals[j];
            if (row.vals[j] > this.MAX_VALUE) {
                this.MAX_VALUE = row.vals[j];
                this.MAX_VALUE_LOG = Math.log(this.MAX_VALUE);
            }

        }
        ;
        if (this.options.cumulative) {
            for (var j = 1; j < this.MAX_UNITS; ++j) {
                values[base_idx + j] += values[base_idx + j - 1];
                if (this.options.cumulative_expires) {
                    for ( var u = 0; u < row.dates_end.length; ++u ) {
                        if ( row.dates_end[u] != null && row.dates_end[u] < (j+1) ) {
                            values[base_idx + j - 1 ] -= row.vals[u]
                        }
                    }
                }
                if (values[base_idx + j] > this.MAX_VALUE) {
                    this.MAX_VALUE = values[base_idx + j];
                    this.MAX_VALUE_LOG = Math.log(this.MAX_VALUE);
                }
            }
        }
    }

    return {
        length:rows.length,
        xcoords:xcoords,
        ycoords:ycoords,
        values:values,
        size:1 << (this.resolution * 2)
    };
};

// get time data in json format
TimePlayer.prototype.get_time_data = function (tile, coord, zoom) {
    var self = this;

    if (!self.table) {
        return;
    }

    // get x, y for cells and sd, se for deforestation changes
    // sd contains the months
    // se contains the deforestation for each entry in sd
    // take se and sd as a matrix [se|sd]
    var numTiles = 1 << zoom;
    var sql = ""
    var column_conv = "";
    var expir_conv = "";

    if (this.options.istime == true){
        column_conv = "date_part('epoch',{0})".format(this.t_column);
        expir_conv = "date_part('epoch',{0})".format(this.options.expiration_column);
    } else {
        column_conv = this.t_column;
        expir_conv = this.options.expiration_column;
    }

    var  z_resolution = function(z) {
      var earth_circumference = 40075017;
      var tile_size = 256;
      var full_resolution = earth_circumference/tile_size;
      return full_resolution / (Math.pow(2,z));
    };

    sql = "WITH cte AS ( " +
            "SELECT ST_SnapToGrid(i.the_geom_webmercator, " +
                                  "CDB_XYZ_Resolution({0})*{1}) g"
                                  . format(zoom, this.resolution) +
            ", {0} c " .format(this.countby) +
            ", floor(({0}- {1})/{2}) d" .format(column_conv, this.MIN_DATE, this.step);

    if ( this.options.cumulative_expires ) {
      sql += ", floor(({0}- {1})/{2}) de".format(expir_conv, this.MIN_DATE, this.step);
    }

    sql += " FROM {0} i\n".format(this.options.table) +
           "WHERE i.the_geom_webmercator && CDB_XYZ_Extent({0}, {1}, {2}) "
              .format(coord.x, coord.y, zoom) +
           " GROUP BY g, d";

    if ( this.options.cumulative_expires ) {
      sql += ", de";
    }

    sql += ") SELECT st_x(g) x, st_y(g) y, array_agg(c) vals, array_agg(d) dates ";

    if ( this.options.cumulative_expires ) {
      sql += ", array_agg(de) dates_end";
    }

    sql += " FROM cte GROUP BY x,y";


    var prof = Profiler.get('tile fetch');
    prof.start();
    this.sql(sql, function (data) {
        if (data.rows) {
            prof.end();
            var p = Profiler.get('tile data cache');
            p.start();
            tile.cells = self.pre_cache_months(data.rows, coord, zoom);
            p.end();
            p = Profiler.get('tile render');
            p.start();
            self.redraw_tile(tile);
            p.end();
        }
    });
};
YO = 1;
TimePlayer.prototype.render_time = function (tile, coord, zoom) {
    var self = this;
    //var month = -this.BASE_UNIT + 1 + this.time>>0;
    //var month = Math.ceil(this.MAX_UNITS * (this.time - this.BASE_UNIT)/(this.CAP_UNIT-this.BASE_UNIT));
    var month = this.time;
    var w = tile.canvas.width;
    var h = tile.canvas.height;
    var ctx = tile.ctx;
    var i, x, y, cell, cells;
    cells = tile.cells;

    if (!cells || cells.length === 0) {
        return;
    }

    var colors = [
        //"#FFFFE5",
        //"#FFF7BC",
        "#FEE391",
        "#FEC44F",
        "#FE9929",
        "#EC7014",
        "#CC4C02",
        "#993404",
        "#662506"
    ];

    var fillStyle;
    // clear canvas    
    tile.canvas.width = w;

    var ci = 0;
    var cu = 0;
    ctx.strokeStyle = ctx.fillStyle = colors[cu];
    ctx.globalCompositeOperation = this.options.blendmode;
    var xc = cells.xcoords;
    var yc = cells.ycoords;
    var vals = cells.values;
    var dz = 256 / Math.pow(2, zoom)

    // render cells
    var len = cells.length;
    var pixel_size = this.resolution//*this.options.cellsize;
    var pixel_size_trail_circ = pixel_size * 2;
    var pixel_size_trail_squa = pixel_size * 1.5;
    var offset = Math.floor((pixel_size - 1) / 2);
    var tau = Math.PI * 2;

    // memoize sprite canvases
    if (self.sprite_1 == undefined) {
        self.sprite_1 = [];
        $(colors).each(function () {
            var canvas = document.createElement('canvas');
            var ctx = canvas.getContext('2d');
            ctx.width = canvas.width = pixel_size * 2;
            ctx.height = canvas.height = pixel_size * 2;
            ctx.globalAlpha = 1;
            ctx.fillStyle = this.toString();
            ctx.beginPath();
            ctx.arc(pixel_size, pixel_size, pixel_size, 0, tau, true, true);
            ctx.closePath();
            ctx.fill();
            self.sprite_1.push(canvas);
        });
    }

    if (self.sprite_2 == undefined) {
        self.sprite_2 = [];
        $(colors).each(function () {
            var canvas = document.createElement('canvas');
            var ctx = canvas.getContext('2d');
            ctx.width = canvas.width = pixel_size_trail_circ * 2;
            ctx.height = canvas.height = pixel_size_trail_circ * 2;
            ctx.globalAlpha = 0.3;
            ctx.fillStyle = this.toString();
            ctx.beginPath();
            ctx.arc(pixel_size_trail_circ, pixel_size_trail_circ, pixel_size_trail_circ, 0, tau, true, true);
            ctx.closePath();
            ctx.fill();
            self.sprite_2.push(canvas);
        });
    }

    if (self.sprite_3 == undefined) {
        self.sprite_3 = [];
        $(colors).each(function () {
            var canvas = document.createElement('canvas');
            var ctx = canvas.getContext('2d');
            ctx.width = canvas.width = pixel_size * 2;
            ctx.height = canvas.height = pixel_size * 2;
            ctx.globalAlpha = 0.3;
            ctx.fillStyle = this.toString();
            ctx.beginPath();
            ctx.arc(pixel_size, pixel_size, pixel_size, 0, tau, true, true);
            ctx.closePath();
            ctx.fill();
            self.sprite_3.push(canvas);
        });
    }


    var numTiles = 1 << zoom;
    for (i = 0; i < len; ++i) {
        var cell = cells.values[this.MAX_UNITS * i + month];
        if (cell) {
            ci = cell == 0 ? 0 : Math.floor((colors.length - 1) * (Math.log(cell) / this.MAX_VALUE_LOG));
            if (ci != cu) {
                cu = ci < colors.length ? ci : cu;
                ctx.fillStyle = colors[cu];
            }
            if (this.options.point_type == 'circle') {
                ctx.drawImage(self.sprite_1[cu], xc[i] - pixel_size, yc[i] - pixel_size)
            } else if (this.options.point_type == 'square') {
                ctx.fillRect(xc[i] - offset, yc[i] - offset, pixel_size, pixel_size);
            }
        }

        if (this.options.trails == true) {

            cell = cells.values[this.MAX_UNITS * i + month - 1];
            if (cell) {
                ci = cell == 0 ? 0 : Math.floor((colors.length - 1) * (Math.log(cell) / this.MAX_VALUE_LOG));
                if (ci != cu) {
                    cu = ci < colors.length ? ci : cu;
                    ctx.fillStyle = colors[cu];
                }
                if (this.options.point_type == 'circle') {
                    //alignment hack - sorry to the gods of graphics
                    ctx.drawImage(self.sprite_2[cu], xc[i] - pixel_size_trail_squa - 1, yc[i] - pixel_size_trail_squa - 1)
                } else if (this.options.point_type == 'square') {
                    ctx.fillRect(xc[i] - offset, yc[i] - offset, pixel_size_trail_squa, pixel_size_trail_squa);
                }
            }

            cell = cells.values[this.MAX_UNITS * i + month - 2];
            if (cell) {
                ci = cell == 0 ? 0 : Math.floor((colors.length - 1) * (Math.log(cell) / this.MAX_VALUE_LOG));
                if (ci != cu) {
                    cu = ci < colors.length ? ci : cu;
                    ctx.fillStyle = colors[cu];
                }
                if (this.options.point_type == 'circle') {
                    ctx.drawImage(self.sprite_3[cu], xc[i] - pixel_size, yc[i] - pixel_size)
                } else if (this.options.point_type == 'square') {
                    ctx.fillRect(xc[i] - offset, yc[i] - offset, pixel_size, pixel_size);
                }
            }
        }
    }
};


/**
 * String formatting for JavaScript.
 *
 * Usage:
 *
 *   "{0} is {1}".format("CartoDB", "epic!");
 *   // CartoDB is epic!
 *
 */
String.prototype.format = (function (i, safe, arg) {
    function format() {
        var str = this,
            len = arguments.length + 1;

        for (i = 0; i < len; arg = arguments[i++]) {
            safe = typeof arg === 'object' ? JSON.stringify(arg) : arg;
            str = str.replace(RegExp('\\{' + (i - 1) + '\\}', 'g'), safe);
        }
        return str;
    }

    //format.native = String.prototype.format;
    return format;
})();


// =================
// profiler
// =================

function Profiler() {
}
Profiler.times = {};
Profiler.new_time = function (type, time) {
    var t = Profiler.times[type] = Profiler.times[type] || {
        max:0,
        min:10000000,
        avg:0,
        total:0,
        count:0
    };

    t.max = Math.max(t.max, time);
    t.total += time;
    t.min = Math.min(t.min, time);
    ++t.count;
    t.avg = t.total / t.count;
};

Profiler.print_stats = function () {
    for (k in Profiler.times) {
        var t = Profiler.times[k];
        console.log(" === " + k + " === ");
        console.log(" max: " + t.max);
        console.log(" min: " + t.min);
        console.log(" avg: " + t.avg);
        console.log(" total: " + t.total);
    }
};

Profiler.get = function (type) {
    return {
        t0:null,
        start:function () {
            this.t0 = new Date().getTime();
        },
        end:function () {
            if (this.t0 !== null) {
                Profiler.new_time(type, this.time = new Date().getTime() - this.t0);
                this.t0 = null;
            }
        }
    };
};
